bundles/org.eclipse.equinox.simpleconfigurator/src/org/eclipse/equinox/internal/simpleconfigurator/ConfigApplier.java

```diff
@@ -252,10 +252,19 @@ private void refreshAllBundles() {
 
 	private Set<Bundle> getDoNotRefresh() {
 		Set<Bundle> doNotRefresh = new HashSet<>();
-		doNotRefresh.add(manipulatingContext.getBundle());
+		Bundle thisBundle = manipulatingContext.getBundle();
+		if (thisBundle != null) {
+			doNotRefresh.add(thisBundle);
+			addDoNotRefreshFragments(thisBundle, doNotRefresh);
```

Comment:
> For example, p2 can manage bundles and set start-levels to 1 for such bundles. I don't think that is a particularly good way to use p2, but it is something that can be done today. In that case why exactly would we not want simple configurator to refresh the bundles that it has installed new versions for. Refreshing such bundles, even if they are started before simple configurator, should never cause simple configurator to be pulled into the refresh closure.

Actually this is what P2 does:

1. create a product with simple configurator
2. add a bundle to the configuration with start level <= simple configurator
3. P2 will add this bundle to the `config.ini` instead of the bundles file of simple configurator

but of course these bundles can be updates, but (from what I can see) this can cause several problems:

Most often, there is a reason for a lower start level, e.g. there are some low-level bundles that should run "always" and might cause other bundles to refresh if they are refreshed (e.g. weaving hooks) what might cause undesired effects, also bundles contain native code, and they are refreshed I sometimes see equinox crash completed with a core-dump. Some singleton bundles also seem not behave "nice" as we can see from wims log file.
Another issue is, that if the framework was entering a lower start-level so that simple configurator was restarted, these other bundles are already restarted as well, so refreshing them will restart them a second time also causing undesired effects.

Of course these can all be updated by P2 but should only be refreshed on next start then as noted in the config.ini, so of course these are very low-level things that should be aware of special states and well aware about how OSGi.

I hope this is a bit clearer now, and not more confusing :-)

Reply:
My opinion is that we should go with this simple fix first to address issue #90.  If incremental improvements are needed for these more advanced scenarios then we can discuss in a separate issue and not lump enhancements related to early start bundles into a fix for issue #90.